### DISCUSSION TITLE
RawRequest is always nil

### DISCUSSION THREAD

**Discussion by konart** [go-resty/resty#470] [Q&A] [answered]
Hi!

I'm trying to use *request.RawRequest in OnBeforeRequest middleware but it's always empty

This is how I'm doing a request (not much going on here)

```go 
        request := SomeStruct{}
	resp, err := resty.R().SetContext(ctx).SetBody(request).Post(urlString)
```

*request.Body has data as expected.

**Comment by jeevatkm** [comment]
@konart Typically `request.RawRequest` will be available for `OnBeforeRequest` middleware. What is the error you're getting as a return value?

Also, can you provide a small test case? it will be helpful to demonstrate an issue.

**Reply by konart** [reply]
Sorry for the (very) late reply!

Here is the example:
```go
package main

import (
	"fmt"
	"github.com/go-resty/resty/v2"
)

func main() {
	// Create a Resty Client
	client := resty.New()

	client.OnBeforeRequest(func(c *resty.Client, req *resty.Request) error {
		if req.RawRequest == nil {
		    fmt.Println("RawRequest is empty")
		}
		return nil
	})

	resp, err := client.R().SetBody(map[string]string{"foo": "bar"}).Post("https://httpbin.org/post")

	// Explore response object
	fmt.Println("Response Info:")
	fmt.Println("  Error      :", err)
	fmt.Println("  Status Code:", resp.StatusCode())
	fmt.Println("  Status     :", resp.Status())
	fmt.Println("  Proto      :", resp.Proto())
	fmt.Println("  Time       :", resp.Time())
	fmt.Println("  Received At:", resp.ReceivedAt())
	fmt.Println("  Body       :\n", resp)
}

```

**Reply by konart** [reply]
Here is the output:
```
RawRequest is empty // <- this is unexpected!
Response Info:
  Error      : <nil>
  Status Code: 200
  Status     : 200 OK
  Proto      : HTTP/2.0
  Time       : 632.017691ms
  Received At: 2021-12-18 11:32:12.674497 +0300 MSK m=+0.632710647
  Body       :
 {
  "args": {},
  "data": "{\"foo\":\"bar\"}",
  "files": {},
  "form": {},
  "headers": {
    "Accept-Encoding": "gzip",
    "Content-Length": "13",
    "Content-Type": "application/json",
    "Host": "httpbin.org",
    "User-Agent": "go-resty/2.7.0 (https://github.com/go-resty/resty)",
    "X-Amzn-Trace-Id": "Root=1-61bd9c8c-0633b031163121af0de7b65d"
  },
  "json": {
    "foo": "bar"
  },
  "origin": "89.189.104.173",
  "url": "https://httpbin.org/post"
}
```

**Reply by konart** [reply]
It seems to me that the issue with this is that user defined middlewares are processed before the internal ones (https://github.com/go-resty/resty/blob/master/client.go#L901) and as a result request constructor `createRequest` (https://github.com/go-resty/resty/blob/master/client.go#L1068) is called after the user's middleware.

**Reply by konart** [reply, marked as answer]
Also see: https://github.com/go-resty/resty/issues/408 and commit https://github.com/go-resty/resty/commit/c6b83faff7284c944f9d5dc66eeaffefe46be591

**Comment by fbury-hec** [comment]
Hi, 

i guess i am having the same issue.
It only applies to POST-Requests.


```
client := r.New()
client.SetDebug(true)
client.OnBeforeRequest(requestLogger)
```

request:

```
resp, err := client.R().
  SetBody(in).
  SetAuthToken(token).
  Post(apiURL + "/data")
```


and middleware:
```
func requestLogger(c *r.Client, req *r.Request) error {
	rr := req.RawRequest
	if rr != nil {
         // always nil for POST, but not for GET
        }
}
```

So i cannot use RawRequest in Middleware.
But `client.SetDebug(true)` is somehow able to access and log RawRequest

**Comment by caojunkai01** [comment]
ping

**Comment by jialechan** [comment]
```go
// import "moul.io/http2curl"
client.SetPreRequestHook(func(client *resty.Client, request *http.Request) error {
    command, _ := http2curl.GetCurlCommand(request)
    fmt.Printf("begin request: %s", command)
    return nil
})

client.OnAfterResponse(func(c *resty.Client, resp *resty.Response) error {
    fmt.Println("response: ", resp)
    return nil
})
```

work for me

**Reply by konart** [reply]
Not sure what's your point here, sorry.

**Comment by robertoej** [comment]
Hi, everyone. We could use the PreHookRequest instead, as mentioned [here](https://github.com/go-resty/resty/pull/355#issuecomment-917816520).

**Comment by slessard** [comment]
I ran into this same problem where `RawRequest` is `nil` thus preventing me from using `SetPreRequestHook`. The root cause for this issue is that I was creating the Resty client incorrectly. 

Wrong way...

```
	var client = &resty.Client{}
```

Correct way...

```
	var client = resty.New()
```

After creating the Resty client the correct way now `RawRequest` has a non-nil value.

**Reply by konart** [reply]
I've been creating it with a constructor method from the beginning though.

As I've said earlier: https://github.com/go-resty/resty/discussions/470#discussioncomment-1840186
